### PR TITLE
OptiGUI 2.3.x rework

### PR DESCRIPTION
**Pull request by opekope2**
## About this version

Fork of OptiGUI 2.1.3. On the 2.1.x branch, there were only bug fixes and compatibility with newer Minecraft versions after this version.

## Short roadmap

* The full OptiGUI 2.2.0 will not be released, because I introduced significant changes, and I didn't like them all, then took half a year off adding new features to OptiGUI. This PR has some of the useful stuff reimplemented, and others will come later.
* Support for older versions (OptiGlue, later LiLaC) were removed. Maintaining compatibility with multiple versions of the game is not a straightforward task, and I no longer have the motivation to do so. As of now, only Minecraft 1.20.4 and later versions will receive OptiGUI 2.3.0, but I'll see how old can I go without the code breaking.
* LiLaC as a standalone mod will get discontinued, and will not receive further updates.

## Developer changes

Basically rewrite the whole API, while making it simpler and much smaller
Also reduce allocation while evaluating texture replacer filters

### CODE REVIEW

**Comment by opekope2** [comment]
Minecraft 1.20.2 changed villager textures, and I don't want to add ifs to that, so Minecraft 1.20.1 and older will no longer be supported.

**Comment by opekope2** [comment]
Fixed #77 
Fixed #84

**Comment by ToppyTopology** [comment]
please reconsider backporting fixes to 1.20.1, as of now it is still regarded as the most popular base version for modern modded minecraft

**Comment by opekope2** [comment]
I only dropped support temporary, while I develop the next full release, because from now on, every incompatible Minecraft version will receive a different JAR instead of using OptiGlue, because it just crashes a lot.
Since the next full release isn't happening this year, I'll backport 2.3.0-beta.5 to Minecraft 1.20.1 later this week.
